---
# title_full: "skater: An R package for SNP-based Kinship Analysis and Benchmarking"
title_full: "skater: An R package for SNP-based Kinship Analysis, Testing, and Evaluation"
title_short: "skater package for kinship analysis"

# One of the following below. See https://academic.oup.com/bioinformatics/pages/instructions_for_authors#Types%20of%20Manuscript
# one of `Original Paper`, `Discovery Note`, `Applications Note`, `Review`, `Letter to the Editor`, `Conference Paper`    
manuscript_type: Applications Note

# One of the following below. See https://academic.oup.com/bioinformatics/pages/instructions_for_authors#Detailed%20Scope%20Guidelines
# Genome analysis
# Sequence analysis
# Phylogenetics
# Structural bioinformatics
# Gene expression
# Genetic and population analysis
# Systems biology
# Data and text mining
# Databases and ontologies
# Bioimage informatics
subject_section:  Genetics and Population Analysis

access_date:
  - day: DD
    month: MM
    year: YY
      
author:
  - name: Stephen D. Turner
    affiliation: "1"
  - name: V.P. Nagraj
    affiliation: "1"
  - name: Matthew Scholz
    affiliation: "1"
  - name: Shakeel Jessa
    affiliation: "1"
  - name: Carlos Acevedo
    affiliation: "1"
  - name: Jianye Ge
    affiliation: "2,3"
  - name: August E. Woerner
    affiliation: "2,3"
  - name: Bruce Budowle
    affiliation: "2,3"

address:
  - code: 1
    address: Signature Science, LLC., Charlottesville, Virginia, 22911, USA.
  - code: 2
    address: Center for Human Identification, University of North Texas Health Science Center, 3500 Camp Bowie Blvd., Fort Worth, TX, 76107, USA
  - code: 3
    address: Department of Microbiology, Immunology, and Genetics, University of North Texas Health Science Center, 3500 Camp Bowie Blvd., Fort Worth, TX, 76107, USA
 
# Full email address to be given, preferably an institution email address.  
corresponding_author_text: "To whom correspondence should be addressed. E-mail: sturner\\@signaturescience.com"
corresponding_author_email: "sturner\\@signaturescience.com"
  
motivation: |
  SNP-based kinship analysis with genome-wide relationship estimation and IBD segment analysis methods produces results that often require further downstream processing and manipulation. A dedicated software package that consistently and intuitively implements this analysis functionality is needed.
  
results: |
  Here we present the skater R package for **S**NP-based **k**inship **a**nalysis, **t**esting, and **e**valuation with **R**. The skater package contains a suite of well-documented tools for importing, parsing, and analyzing pedigree data, performing relationship degree inference, benchmarking relationship degree classification, and summarizing IBD segment data.

availability: |
  The skater package is implemented as an R package and is released under the MIT license at https\:\/\/github.com/signaturescience/skater. Documentation is available at https\:\/\/signaturescience.github.io/skater.

supplementary_info: |
  Supplementary data are available at Bioinformatics Online.

output:
  rticles::bioinformatics_article:
    citation_package: natbib
    
bibliography: bibliography.bib
---

```{r, echo=FALSE, eval=FALSE}
## Run this chunk interactively if you update the package vignette.
## This renders the vignette as a PDF with a new title.

# Where is the package vignette?
vpath <- here::here("vignettes/basic_usage.Rmd")
# Where do you want to save the new modified vignette rmarkdown?
spath <- here::here("paper/supplement-vignette.Rmd")
# Read in the vignette rmarkdown text
vtext <- readLines(vpath)
# Create the supplemental material rmarkdown by replacing the title YAML
stext <- sub(pattern='title: "Basic Usage"', 
             replacement = 'title: |\n    | Supplementary Material for:\n    | skater: An R package for SNP-based Kinship Analysis, Testing, and Evaluation\n    | _(skater v0.1.0 package vignette)_', 
             x=vtext)
# Write out that modified yaml to the supplementary material rmarkdown
writeLines(text=stext, con=spath)
# Render it to PDF
rmarkdown::render(spath, output_format="pdf_document")
```


# Introduction

Inferring familial relationships between individuals using genetic data is a common practice in population genetics, medical genetics, and forensics. There are multiple approaches to estimating relatedness between samples, including genome-wide measures, such as those implemented in Plink [@purcell2007] or KING [@manichaikul2010], and methods that rely on identity by descent (IBD) segment detection, such as GERMLINE [@gusev2009], hap-IBD [@zhou2020], and IBIS [@seidman2020]. Recent efforts focusing on benchmarking these methods [@ramstetter2017; @vries2021] have been aided by tools for simulating pedigrees and genome-wide SNP data [@caballero2019]. Analyzing results from genome-wide SNP-based kinship analysis or comparing analyses to simulated data for benchmarking have to this point required writing one-off analysis functions or utility scripts that are seldom distributed with robust documentation, test suites, or narrative examples of usage. There is a need in the field for a well-documented software package with a consistent design and API that contains functions to assist with downstream manipulation, benchmarking, and analysis of SNP-based kinship assessment methods. Here we present the skater package for **S**NP-based **k**inship **a**nalysis, **t**esting, and **e**valuation with **R**.

# The skater package

The skater package provides an intuitive collection of analysis and utility functions for SNP-based kinship analysis. Functions in the package include tools for importing, parsing, and analyzing pedigree data, performing relationship degree inference, benchmarking relationship degree classification, and summarizing IBD segment data. The package is designed to adhere to "tidy" data analysis principles, and builds upon the tools released under the tidyverse R ecosystem [@Wickham2019].

## Pedigree parsing, manipulation, and analysis

The skater package has several functions for importing, parsing, and analyzing pedigree data. Pedigrees define familial relationships in a hierarchical structure. Many genomics tools for working with pedigrees start with a .fam file, which is a tabular format with one row per individual and columns for unique IDs of the mother, father, and the family unit. The skater package contains the function `read_fam()` to read in a PLINK-formatted .fam file and another function `fam2ped()` to convert the content into a pedigree object as a nested tibble with one row per family. All pedigree processing from skater internally leverages a data structure from the kinship2 package [@kinship2]. Further functions such as `plot_pedigree()` produce a multi-page PDF drawing a diagram of the pedigree for each family, while `ped2kinpair()` produces a pairwise list of relationships between all individuals in the data with the expected kinship coefficients for each pair (see Supplementary Material).

## Relationship degree inference and benchmarking

The skater package includes functions to translate kinship coefficients to relationship degrees. The kinship coefficients could come from `ped2kinpair()` or other kinship estimation software.

The `dibble()` function creates a **d**egree **i**nference t**ibble**, with degrees up to the specified maximum degree resolution, expected kinship coefficient, and lower and upper inference ranges as defined in @manichaikul2010. The `kin2degree()` function infers the relationship degree given a kinship coefficient and a maximum degree resolution (e.g., 7th-degree relatives) up to which anything more distant is classified as unrelated.

Once estimated kinship is converted to degree, it may be of interest to compare the inferred degree to known degrees of relatedness. When aggregated over many relationships and inferences, this can help benchmark performance of a particular kinship analysis method. The skater package adapts a `confusion_matrix()` function from @clark2021 to provide standard contingency table metrics (e.g. sensitivity, specificity, PPV, precision, recall, F1, etc.) with a new reciprocal RMSE (R-RMSE) metric. The R-RMSE metric is defined more thoroughly in the skater package vignette (see Supplementary Material) and may be a preferable measure of classification accuracy when benchmarking relationship degree estimation. Taking the reciprocal of the target and predicted degree in a typical RMSE calculation results in larger penalties for more egregious misclassifications (e.g., classifying a first-degree relative pair as second-degree) than misclassifications at more distant relationships (e.g., classifying a fourth-degree relative pair as fifth-degree).

## IBD segment analysis

Tools such as hap-IBD [@zhou2020], and IBIS [@seidman2020] detect shared IBD segments between individuals. The skater package includes functionality to take those IBD segments, compute shared genomic centimorgan (cM) length, and converts that shared cM to a kinship coefficient. In addition to inferred segments, these functions can estimate "truth" kinship from simulated IBD segments [@caballero2019]. The `read_ibd()` function reads pairwise IBD segments from IBD inference tools and from simulated IBD segments. The `read_map()` function reads in genetic map in a standard format which is required to translate the total centimorgans shared IBD to a kinship coefficient using the `ibd2kin()` function.

# Conclusion

The skater R package provides a robust software package for data import, manipulation, and analysis tasks typically encountered when working with SNP-based kinship analysis tools. All package functions are internally documented with examples, and the package contains a vignette demonstrating usage, inputs, outputs, and interpretation of all key functions (see Supplementary Material). The package contains internal tests that are automatically run with continuous integration via GitHub Actions whenever the package code is updated. The skater package is permissively licensed (MIT) and is easily extensible to accommodate outputs from new genome-wide relatedness and IBD segment methods as they become available.

<!-- # Acknowledgements {.unnumbered} -->

<!-- These should be included at the end of the text and not in footnotes. Please ensure you acknowledge all sources of funding, see funding section below. -->

<!-- Details of all funding sources for the work in question should be given in a separate section entitled 'Funding'. This should appear before the 'Acknowledgements' section. -->

# Funding {.unnumbered}

This work was supported in part by award 2019-DU-BX-0046 (Dense DNA Data for Enhanced Missing Persons Identification) to B.B., awarded by the National Institute of Justice, Office of Justice Programs, U.S. Department of Justice and by internal funds from the Center for Human Identification. The opinions, findings, and conclusions
or recommendations expressed are those of the authors and do not necessarily reflect those of the U.S. Department of Justice.

<!-- -   The sentence should begin: 'This work was supported by ...' - -->
<!-- -   The full official funding agency name should be given, i.e. 'National Institutes of Health', not 'NIH' (full RIN-approved list of UK funding agencies) -->
<!-- -   Grant numbers should be given in brackets as follows: '[grant number xxxx]' -->
<!-- -   Multiple grant numbers should be separated by a comma as follows: '[grant numbers xxxx, yyyy]' -->
<!-- -   Agencies should be separated by a semi-colon (plus 'and' before the last funding agency) -->
<!-- -   Where individuals need to be specified for certain sources of funding the following text should be added after the relevant agency or grant number 'to [author initials]'. -->

<!-- An example is given here: 'This work was supported by the National Institutes of Health [AA123456 to C.S., BB765432 to M.H.]; and the Alcohol & Education Research Council [hfygr667789].' -->

<!-- Oxford Journals will deposit all NIH-funded articles in PubMed Central. See Depositing articles in repositories -- information for authors for details. Authors must ensure that manuscripts are clearly indicated as NIH-funded using the guidelines above. -->
