\documentclass{bioinfo}
\copyrightyear{2016} \pubyear{2016}

%% Some pieces required from the pandoc template
\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}


% Pandoc citation processing


% hyperref makes the margins screwy.
% https://groups.google.com/forum/#!topic/latexusersgroup/4W_SwGk6zx4
% http://ansuz.sooke.bc.ca/software/latex-tricks.php
% \usepackage[colorlinks=true, allcolors=blue]{hyperref}

\access{Advance Access Publication Date:   }
\appnotes{Applications Note}

\begin{document}
\firstpage{1}

\subtitle{Genetics and Population Analysis}

\title[skater package for kinship analysis]{skater: An R package for
SNP-based Kinship Analysis, Testing, and Evaluation}

\author[FirstAuthorLastName \textit{et~al}.]{
Stephen D. Turner\,\textsuperscript{1}, V.P.
Nagraj\,\textsuperscript{1}, Matthew
Scholz\,\textsuperscript{1}, Shakeel Jessa\,\textsuperscript{1}, Carlos
Acevedo\,\textsuperscript{1}, Jianye Ge\,\textsuperscript{2,3}, August
Woerner\,\textsuperscript{2,3}, Bruce Budowle\,\textsuperscript{2,3}
}

\address{
\textsuperscript{1}Signature Science, LLC., Charlottesville, Virginia,
22911, USA.\\
\textsuperscript{2}Center for Human Identification, University of North
Texas Health Science Center, 3500 Camp Bowie Blvd., Fort Worth, TX,
76107, USA\\
\textsuperscript{3}Department of Microbiology, Immunology, and Genetics,
University of North Texas Health Science Center, 3500 Camp Bowie Blvd.,
Fort Worth, TX, 76107, USA\\
}

\corresp{To whom correspondence should be addressed. E-mail:
sturner@signaturescience.com}

\history{Received on XXX; revised on XXX; accepted on XXX}

\editor{Associate Editor: XXX}

\abstract{
\textbf{Motivation:} SNP-based kinship analysis with genome-wide
relationship estimation and IBD segment analysis methods produces
results that often require further downstream processing and
manipulation. A dedicated software package that consistently and
intuitively implements this analysis functionality is needed.\\
\textbf{Results:} Here we present the skater R package for
\textbf{S}NP-based \textbf{k}inship \textbf{a}nalysis, \textbf{t}esting,
and \textbf{e}valuation with \textbf{R}. The skater package contains a
suite of well-documented tools for importing, parsing, and analyzing
pedigree data, performing relationship degree inference, benchmarking
relationship degree classification, and summarizing IBD segment data.\\
\textbf{Availability:} The skater package is implemented as an R package
and is released under the MIT license at
https://github.com/signaturescience/skater. Documentation is available
at https://signaturescience.github.io/skater.\\
\textbf{Contact:}sturner@signaturescience.com\\
\textbf{Supplementary information:} Supplementary data are available at
Bioinformatics Online.}

\maketitle

\section{Introduction}

Inferring familial relationships between individuals using genetic data
is a common practice in population genetics, medical genetics, and
forensics. There are multiple approaches to estimating relatedness
between samples, including genome-wide measures, such as those
implemented in Plink \citep{purcell2007} or KING
\citep{manichaikul2010}, and methods that rely on identity by descent
(IBD) segment detection, such as GERMLINE \citep{gusev2009}, hap-IBD
\citep{zhou2020}, and IBIS \citep{seidman2020}. Recent efforts focusing
on benchmarking these methods \citep{ramstetter2017, vries2021} have
been aided by tools for simulating pedigrees and genome-wide SNP data
\citep{caballero2019}. Analyzing results from genome-wide SNP-based
kinship analysis or comparing analyses to simulated data for
benchmarking have to this point required writing one-off analysis
functions or utility scripts that are seldom distributed with robust
documentation, test suites, or narrative examples of usage. There is a
need in the field for a well-documented software package with a
consistent design and API that contains functions to assist with
downstream manipulation, benchmarking, and analysis of SNP-based kinship
assessment methods. Here we present the skater package for
\textbf{S}NP-based \textbf{k}inship \textbf{a}nalysis, \textbf{t}esting,
and \textbf{e}valuation with \textbf{R}.

\section{The skater package}

The skater package provides an intuitive collection of analysis and
utility functions for SNP-based kinship analysis. Functions in the
package include tools for importing, parsing, and analyzing pedigree
data, performing relationship degree inference, benchmarking
relationship degree classification, and summarizing IBD segment data.
The package is designed to adhere to ``tidy'' data analysis principles,
and builds upon the tools released under the tidyverse R ecosystem
\citep{Wickham2019}.

\subsection{Pedigree parsing, manipulation, and analysis}

The skater package has several functions for importing, parsing, and
analyzing pedigree data. Pedigrees define familial relationships in a
hierarchical structure. Many genomics tools for working with pedigrees
start with a .fam file, which is a tabular format with one row per
individual and columns for unique IDs of the mother, father, and the
family unit. The skater package contains the function
\texttt{read\_fam()} to read in a PLINK-formatted .fam file and another
function \texttt{fam2ped()} to convert the content into a pedigree
object as a nested tibble with one row per family. All pedigree
processing from skater internally leverages a data structure from the
kinship2 package \citep{kinship2}. Further functions such as
\texttt{plot\_pedigree()} produce a multi-page PDF drawing a diagram of
the pedigree for each family, while \texttt{ped2kinpair()} produces a
pairwise list of relationships between all individuals in the data with
the expected kinship coefficients for each pair.

\subsection{Relationship degree inference and benchmarking}

The skater package includes functions to translate kinship coefficients
to relationship degrees. The kinship coefficients could come from
\texttt{ped2kinpair()} or other kinship estimation software.

The \texttt{dibble()} function creates a \textbf{d}egree
\textbf{i}nference t\textbf{ibble}, with degrees up to the specified
maximum degree resolution, expected kinship coefficient, and lower and
upper inference ranges as defined in \citet{manichaikul2010}. The
\texttt{kin2degree()} function infers the relationship degree given a
kinship coefficient and a maximum degree resolution (e.g., 7th-degree
relatives) up to which anything more distant is classified as unrelated.

Once estimated kinship is converted to degree, it may be of interest to
compare the inferred degree to known degrees of relatedness. When
aggregated over many relationships and inferences, this can help
benchmark performance of a particular kinship analysis method. The
skater package adapts a \texttt{confusion\_matrix()} function from
\citet{clark2021} to provide standard contingency table metrics
(e.g.~sensitivity, specificity, PPV, precision, recall, F1, etc.) with a
new reciprocal RMSE (R-RMSE) metric. The R-RMSE metric is defined more
thoroughly in the skater package vignette (see Supplementary Material)
and may be a preferable measure of classification accuracy when
benchmarking relationship degree estimation. Taking the reciprocal of
the target and predicted degree in a typical RMSE calculation results in
larger penalties for more egregious misclassifications (e.g.,
classifying a first-degree relative pair as second-degree) than
misclassifications at more distant relationships (e.g., classifying a
fourth-degree relative pair as fifth-degree).

\subsection{IBD segment analysis}

Tools such as hap-IBD \citep{zhou2020}, and IBIS \citep{seidman2020}
detect shared IBD segments between individuals. The skater package
includes functionality to take those IBD segments, compute shared
genomic centimorgan (cM) length, and converts that shared cM to a
kinship coefficient. In addition to inferred segments, these functions
can estimate ``truth'' kinship from simulated IBD segments
\citep{caballero2019}. The \texttt{read\_ibd()} function reads pairwise
IBD segments from IBD inference tools and from simulated IBD segments.
The \texttt{read\_map()} function reads in genetic map in a standard
format which is required to translate the total centimorgans shared IBD
to a kinship coefficient using the \texttt{ibd2kin()} function.

\section{Conclusion}

The skater R package provides a robust software package for data import,
manipulation, and analysis tasks typically encountered when working with
SNP-based kinship analysis tools. All package functions are internally
documented with examples, and the package contains a vignette
demonstrating usage, inputs, outputs, and interpretation of all key
functions (see Supplementary Material). The package contains internal
tests that are automatically run with continuous integration via GitHub
Actions whenever the package code is updated. The skater package is
permissively licensed (MIT) and is easily extensible to accommodate
outputs from new genome-wide relatedness and IBD segment methods as they
become available.

\section*{Funding}
\addcontentsline{toc}{section}{Funding}

This work was supported in part by award 2019-DU-BX-0046 (Dense DNA Data
for Enhanced Missing Persons Identification) to B.B., awarded by the
National Institute of Justice, Office of Justice Programs, U.S.
Department of Justice and by internal funds from the Center for Human
Identification. The opinions, findings, and conclusions or
recommendations expressed are those of the authors and do not
necessarily reflect those of the U.S. Department of Justice.


% Bibliography
\bibliographystyle{natbib}
\bibliography{bibliography.bib}

\end{document}
